aws/resource_aws_athena_database.go

```diff
@@ -0,0 +1,208 @@
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceAwsAthenaDatabaseCreate(d *schema.ResourceData, meta interface{}) error {
+	athenaconn := meta.(*AWSClient).athenaconn
+
+	athenainput := &athena.StartQueryExecutionInput{
+		QueryString: aws.String(createDatabaseQueryString(d.Get("name").(string))),
+		ResultConfiguration: &athena.ResultConfiguration{
+			OutputLocation: aws.String("s3://" + d.Get("bucket").(string)),
+		},
+	}
+
+	athenaresp, err := athenaconn.StartQueryExecution(athenainput)
+	if err != nil {
+		return err
+	}
+
+	if err := checkCreateDatabaseQueryExecution(*athenaresp.QueryExecutionId, d, meta); err != nil {
+		return err
+	}
+	d.SetId(d.Get("name").(string))
+	return resourceAwsAthenaDatabaseRead(d, meta)
+}
+
+func resourceAwsAthenaDatabaseRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).athenaconn
+
+	bucket := d.Get("bucket").(string)
+	input := &athena.StartQueryExecutionInput{
+		QueryString: aws.String(showDatabaseQueryString()),
+		ResultConfiguration: &athena.ResultConfiguration{
+			OutputLocation: aws.String("s3://" + bucket),
+		},
+	}
+
+	resp, err := conn.StartQueryExecution(input)
+	if err != nil {
+		return err
+	}
+
+	if err := checkShowDatabaseQueryExecution(*resp.QueryExecutionId, d, meta); err != nil {
+		return err
+	}
+	return nil
+}
+
+func resourceAwsAthenaDatabaseUpdate(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+
+func resourceAwsAthenaDatabaseDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).athenaconn
+
+	bucket := d.Get("bucket").(string)
+	input := &athena.StartQueryExecutionInput{
+		QueryString: aws.String(dropDatabaseQueryString(d.Get("name").(string))),
+		ResultConfiguration: &athena.ResultConfiguration{
+			OutputLocation: aws.String("s3://" + bucket),
+		},
+	}
+
+	resp, err := conn.StartQueryExecution(input)
+	if err != nil {
+		return err
+	}
+
+	if err := checkDropDatabaseQueryExecution(*resp.QueryExecutionId, d, meta); err != nil {
+		return err
+	}
+	return nil
+}
+
+func createDatabaseQueryString(databaseName string) string {
+	return fmt.Sprintf("create database %s;", databaseName)
+}
+
+func checkCreateDatabaseQueryExecution(qeid string, d *schema.ResourceData, meta interface{}) error {
+	rs, err := queryExecutionResult(qeid, meta)
+	if err != nil {
+		return err
+	}
+	if len(rs.Rows) != 0 {
+		return fmt.Errorf("[ERROR] Athena create database, unexpected query result: %s", flattenAthenaResultSet(rs))
+	}
+	return nil
+}
+
+func showDatabaseQueryString() string {
+	return fmt.Sprint("show databases;")
+}
+
+func checkShowDatabaseQueryExecution(qeid string, d *schema.ResourceData, meta interface{}) error {
+	rs, err := queryExecutionResult(qeid, meta)
+	if err != nil {
+		return err
+	}
+	found := false
+	dbName := d.Get("name").(string)
+	for _, row := range rs.Rows {
+		for _, datum := range row.Data {
+			if *datum.VarCharValue == dbName {
+				found = true
+			}
+		}
+	}
```

Comment:
Can't we just always expect 1 row to be returned and error out if that's not the case? Instead of nesting the for loops here?

Reply:
I depend on that `show databases;` won't return the same name databases.
So if the query return multiple same database, this function doesn't error out

Reply:
I agree that this operation isn't clear 😕